src/lavinmq/vhost.cr

```diff
@@ -611,15 +614,24 @@ module LavinMQ
         @replicator.replace_file @definitions_file_path
         @definitions_file.close
         @definitions_file = io
+        @definitions_dirt_counter = 0
       end
     end
 
-    private def store_definition(frame)
+    private def store_definition(frame, dirty = false)
       @log.debug { "Storing definition: #{frame.inspect}" }
       bytes = frame.to_slice
       @definitions_file.write bytes
       @replicator.append @definitions_file_path, bytes
       @definitions_file.fsync
+      if dirty
+        @definitions_dirt_counter += 1
+        # By doing equality comparision we'll only start one fiber
+        # without having to keep track if it's started or not
+        if @definitions_dirt_counter == DEFINITIONS_DIRT_COMPACT_THREASHOLD
+          spawn(name: "VHost #{name} definitions compact") { compact! }
```

Comment:
why in a separate fiber? makes it potentially thread unsafe

Reply:
I think one reason was that `store_definition` is called from a  `@definitions_lock.synchronize` context, and that `compact!` also want the lock, and the other was to not block the current fiber with this (even though it's probably fast, and it will block by holding the lock).